### DISCUSSION TITLE
Support tuple pattern matching for determined assignment

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7357] [Language Ideas] [answered]
**EDIT**: This is not a bug, see the answer.

----

Currently, the following pattern recognizes the definite assignment of the captured variable in the pattern:
```csharp
if (x is not DirectoryInfo xf)
{
    // code, perhaps return or throw
}

xf.Delete(); // can be used here safely
```

But using tuple pattern matching breaks the recognition:
```csharp
if ((x, y) is not (DirectoryInfo xf, DirectoryInfo yf))
{
    throw new ArgumentException("Can only handle DirectoryInfo instances");
}

// Neither can be used
xf.Delete();
yf.Delete();
```

Logically, the tuple matching expression would always have `xf` and `yf` assigned outside the scope of the `if` statement.

The general requirements from the above example is:
> when the `is not` pattern refers to a tuple capturing variables, those captured variables should be usable outside the `if` statement's scope

**Comment by 333fred** [comment]
This just seems like it's a bug. Consider opening on roslyn.

**Reply by CyrusNajmabadi** [reply]
@jcouv can you weigh in on if thsi is an impl bug, or if things are working as intended here?  Paging you due to your experience with tuples.  If someone else is better suited to answer this, def delegate :)

**Comment by bernd5** [comment, marked as answer]
The code actually runs fine. Which version do you use?
See: [here](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYB8ACAmAjAWAChMAGAAk1ygG4iiB7YAKwFMBjOMsMgXjNWQAbQbUKNWHMgE9e/ISKIBLAGZkAFGrAxpASjKKAzv3qc1AGUNwAPItRwAfF2XaLB67YfTlOnUQDeRGRBZHAAFoj0AO78LNEAgogA5sgAtix2AKJgbCwADnCK9KhqAEQAwgCGqGRFgjKhVQAmgixkACKKiOxw9IhSAJKoyvT6qG5VOQYlOqIAvnSEAPSLZAByLIphLIhkbFVkwK3IBiyNRJQAnJrKAHRl9GhwM+e4V1K394/PhEA===)

(I simply changed the type - DirectoryInfo is not allowed in SharpLab..)

**Reply by Rekkonnect** [reply]
I confirm that both examples work in Visual Studio 2022 version 17.6.5, but this is on my work computer. Last night I was doing that at home and discovered that it failed to compile, but I'm not sure what could be causing this problem. Perhaps it was a temporary regression, will update with more data.

**Reply by Rekkonnect** [reply]
Update: this actually works, I have no idea how I thought it was wrong, perhaps forgot to type something. This is not a bug.
